src/main/scala/dpla/ingestion3/utils/OutputHelper.scala

```diff
@@ -1,38 +1,173 @@
 package dpla.ingestion3.utils
 
-import java.time.LocalDateTime
 import java.time.format.DateTimeFormatter
+import java.time.LocalDateTime
+import java.io.{BufferedWriter, File, FileWriter}
+
+import com.amazonaws.services.s3.model.{PutObjectRequest, PutObjectResult}
+import com.amazonaws.services.s3.AmazonS3Client
+
+import scala.util.Try
 
-object OutputHelper {
+/*
+ * @param root: Root directory or AWS S3 bucket.
+ *              If `root' is an AWS S3 bucket, it should start with "s3a://".
+ * @param shortName: Provider short name
+ * @param activity: "harvest", "mapping", "enrichment", etc.
+ * @param startDateTime: start dateTime of the activity
+ *
+ * @throws IllegalArgumentException
+ *
+ * @see https://digitalpubliclibraryofamerica.atlassian.net/wiki/spaces/TECH/pages/84512319/Ingestion+3+Storage+Specification
+ *      for details on file naming conventions
+ *
+ */
+class OutputHelper(root: String,
+                   shortName: String,
+                   activity: String,
+                   startDateTime: LocalDateTime) {
 
   /*
-   * Get output path for a harvest, mapping, enrichment, or indexing activity.
-   *
-   * @example:
-   *   outputPath("s3://dpla-master-dataset", "cdl", "harvest") =>
-   *   "s3://dpla-master-dataset/cdl/harvest/20170209_104428-cdl-OriginalRecord.avro"
+   * If root is an S3 bucket, ensure that s3a protocol is being used and not
+   * s3 or s3n.
+   */
+  require(!root.startsWith("s3://") && !root.startsWith("s3n://"),
+    "s3a protocol required for writing output")
+
+  /*
+   * File name for a harvest, mapping, enrichment, indexing (etc.) activity.
+   * For full output path, including root directory/bucket, use `outputPath'
+   * Does not include starting "/"
    *
-   * @see https://digitalpubliclibraryofamerica.atlassian.net/wiki/spaces/TECH/pages/84512319/Ingestion+3+Storage+Specification
+   * Evaluate on instantiation so that invalid `activity' is caught immediately.
    */
-  def outputPath(directory: String, shortName: String, activity: String): String = {
+  val fileKey: String = {
 
     val formatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")
-    val timestamp: String = LocalDateTime.now.format(formatter)
+
+    val timestamp: String = startDateTime.format(formatter)
 
     // TODO: handle "reports" case
     // TODO: handle "logs" case
-    val schema = activity match {
+    // TODO: make schema configurable?
+    val schema: String = activity match {
       case "harvest" => "OriginalRecord"
```

Comment:
This list of operations looks potentially like a good place to use a sealed case class hierarchy rather than strings. Then the case clasess could store things like the schema rather than having to spread this type of information around the project.